lib/authenticate_methods.rb

```diff
@@ -50,35 +50,35 @@ def auto_login_external_user
     # Try to to find an existing user who has a matching external UUID
     # which indicates the account has been linked before
     # if found, we trust it and sign the user in automatically
-    if incoming_user.present?
-      sign_out!(security_log_data: {type: 'new external user'}) if signed_in?
-      sign_in!(incoming_user, security_log_data: {type: 'external uuid'})
-      return true
+    incoming_user = UserExternalUuid.find_by_uuid(external_user_uuid).try(:user)
+    return false if incoming_user.nil?
+
+    if signed_in? && incoming_user != current_user
+      # Sign out the current user; signing in the new user will effectively sign out
+      # the current user, but we choose to sign out explicitly so we can record
+      # information in the current user's security log
+
+      sign_out!(security_log_data: {type: 'different external user'})
     end
-    false
+
+    sign_in!(incoming_user, security_log_data: {type: 'external uuid'})
+    true
   end
 
-  def prepare_new_external_user_signup
+  def prepare_for_new_external_user
     # If we didn't find a user with a linked account to automatically log in,
-    # we do not want to assume that any already-logged-in user owns this secure
+    # we do not want to assume that any already-logged-in user owns this signed
     # params information.
     # Therefore at this point we sign out whoever is signed in.
-    if signed_in?
-      type = incoming_user == current_user ?
-               'untrusted external user' : 'different external user'
-      sign_out!(security_log_data: { type: type })
-    end
```

Comment:
if the incoming user is the current user, I don't think we should sign them out, but I'm open to hearing the thoughts behind `untrusted external user`.